pyartnet/dmx_client.py

```diff
@@ -0,0 +1,229 @@
+                if out.endswith('...'):
+                    continue
+
+                out_desc += '  - '
+                out += ' ...'
+                continue
+
+            # format block of channels
+            _block_vals = []
+            _block_desc = []
+            for i in range(5):
+                if k + i < _max_channel:
+                    if show_description:
+                        _block_desc.append(f'{k + i + 1:<3d}')
+                    _block_vals.append(f'{p[18 + k + i]:03d}')
+
+            # separator
+            if out.endswith('...'):
+                out_desc += ' '
+                out += ' '
+            else:
+                out_desc += '   '
+                out += '   '
+
+            out += ' '.join(_block_vals)
+            if show_description:
+                out_desc += ' '.join(_block_desc)
+
+        if show_description:
+            log.debug(out_desc)
+        log.debug(out)
+
+
+class KiNetClient(DmxClient):
+    """
+    Interface with a KiNet device
+    """
+
+    def __init__(self, host: str, port: int):
+        """
+        :param host: IP of the Art-Net Node
+        :param port: Port of the Art-Net Node
+        """
+        super().__init__(host, port)
+
+        self.__socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # UDP
+
+        packet = bytearray()
+        packet.extend(struct.pack(">IHH", 0x0401DC4A, 0x0100, 0x0101))  # Magic, version, type
+        packet.extend(
+            struct.pack(">IBBHI", 0, 0, 0, 0, 0xFFFFFFFF)
+        )  # sequence, port, padding, flags, timer
+        self.__base_packet = packet
+
+    def update(self, universe_nr, universe):
+        """
+        Send the current state of DMX values to the gateway via UDP packet.
+        """
+        packet = self.__base_packet[:]
+        packet.extend(struct.pack("B", universe_nr))  # Universe
+        packet.extend(universe.data)
+        self.__socket.sendto(packet, (self.__host, self.__port))
+        log.debug(f"Sending Art-Net frame to {self.__host}:{self.__port}")
+
+
+class SacnClient(DmxClient):
+    """
+    Interface with a sACN device
+    """
+
+    def __init__(self, host: str, port: int):
+        """
+        :param host: IP of the Art-Net Node
+        :param port: Port of the Art-Net Node
+        """
+        super().__init__(host, port)
+
+        # Initialise socket
+        self._socket = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)  # UDP
+
+        packet = bytearray()
+        # Root layer
+        packet.extend([0x00, 0x01, 0x00, 0x00])  # Preamble Size, Post-amble Size
+        packet.extend(map(ord, "ASC-E1.17"))  # Packet Identifier
+        packet.extend([0x00, 0x00, 0x00, 0x72, 0x57])  # padding x 3 , Flags, Length
+        packet.extend(struct.pack(">l", 4))  # Root Layer Vector
+        packet.extend(map(ord, "ThisIsMyCIDxxxxx"))  # CID, a unique identifier
+        # Framing layer
+        packet.extend([0x72, 0x57])  # Flags and length
```

Comment:
Why is the length set to a static value? Shouldn't it be dynamic based on the highest channel?

Reply:
That's only the length of the Framing layer, which is constant. The actual size of the universe data is hardcoded on line 224. (` packet.extend([0x72, 0x0d,`)

The synchronization mechanism of the E1.31 protocol seems to send the entire universe at once, so no highest channel can be selected as far as I've seen.

[ANSI_E1-31-2018.pdf](https://github.com/spacemanspiff2007/PyArtNet/files/9131501/ANSI_E1-31-2018.pdf)